core-addon/DataCollection.js

```diff
@@ -0,0 +1,128 @@
+   *
+   * @param {String} [studyAddonid=undefined]
+   *        optional study id. It's sent in the ping, if present, to signal
+   *        that user enroled in the study.
+   */
+  async sendEnrollmentPing(studyAddonId) {
+    // If we were provided with a study id, then this is an enrollment to a study.
+    // Send the id alongside with the data and change the schema namespace to simplify
+    // the work on the ingestion pipeline.
+    if (typeof studyAddonId != "undefined") {
+      return await this._sendEmptyPing("pioneer-enrollment", studyAddonId);
+    }
+
+    // Note that the schema namespace directly informs how data is segregated after ingestion.
+    // If this is an enrollment ping for the pioneer program (in contrast to the enrollment to
+    // a specific study), use a meta namespace.
+    return await this._sendEmptyPing("pioneer-enrollment", "pioneer-meta");
+  }
+
+  /**
+   * Sends a Ion deletion-request ping.
+   *
+   * @param {String} studyAddonid
+   *        It's sent in the ping to signal that user unenrolled from a study.
+   */
+  async sendDeletionPing(studyAddonId) {
+    if (typeof studyAddonId === undefined) {
+      throw new Error("IonCore - the deletion-request ping requires a study id");
+    }
+
+    return await this._sendEmptyPing("deletion-request", studyAddonId);
+  }
+
+  /**
+   * Send a ping using the Firefox legacy telemetry.
+   *
+   * @param {String} payloadType
+   *        The type of the encrypted payload. This will define the
+   *        `schemaName` of the ping.
+   * @param {Object} payload
+   *        A JSON-serializable payload to be sent with the ping.
+   * @param {String} namespace
+   *        The namespace to route the ping. This will define the
+   *        `schemaNamespace` and `studyName` properties of the ping.
+   * @param {String} keyId
+   *        The id of the key used to encrypt the payload.
+   * @param {Object} key
+   *        The key used to encrypt the payload, for example:
+   *        {
+   *          crv: "P-256", kty: "EC",
+   *          x: "XLkI3NaY3-AF2nRMspC63BT1u0Y3moXYSfss7VuQ0mk",
+   *          y: "SB0KnIW-pqk85OIEYZenoNkEyOOp5GeWQhS1KeRtEUE",
+   *        }
+   */
+  async sendPing(payloadType, payload, namespace, keyId, key) {
+    let options = {
+      studyName: namespace,
+      addPioneerId: true,
+      encryptionKeyId: keyId,
+      publicKey: key,
+      schemaName: payloadType,
+      schemaVersion: 1,
+      // Note that the schema namespace directly informs how data is
+      // segregated after ingestion.
+      schemaNamespace: namespace,
+    };
+
+    // We intentionally don't wait on the promise returned by
+    // `submitExternalPing`, because that's an internal API only meant
+    // for telemetry tests. Moreover, in order to send a custom schema
+    // name and a custom namespace, we need to ship a custom "experimental"
+    // telemetry API for legacy telemetry.
```

Comment:
We *could* change the `browser.telemetry.submitEncryptedPing` API so that it behaves the way we want here. I think whether that is worth it depends on how long we expect glean.js to take :)

Reply:
The fewer custom APIs we ship in the add-on the better, since there's actual test coverage in-tree for `browser.telemetry` so we won't be caught by surprise if there is some refactoring to the built-in telemetry in Firefox.

Reply:
This is a good point! Note, though, that `submitEncryptedPing` is basically sugar on the top of `TelemetryController.submitExternalPing`, which is a stable API surface. I would hardly believe that's going to change anytime soon! But, even if it does, the telemetry team (of which I'm part ;-) ) usually sends a PSA email long before that happens.

I'd recommend sticking to the current plan and tackling this if we have urgent needs as this requires trains-work!